File: bindings/nodejs/test/kdf.test.js
'use strict';

/**
 * `crypto_kdf` - no official vector exists for this construction (D-45: no DSTU KDF standard or
 * reference implementation exists at all). Correctness here means determinism/distinctness,
 * matching the Rust crate's own property-test posture. Misuse: wrong-length master key/context,
 * negative subkeyId (this binding's own JS-boundary check, D-126 - no rejection category, there
 * is no tag to tamper with). Mirrors bindings/python/tests/test_kdf.py.
 */

const test = require('node:test');
const assert = require('node:assert/strict');
const dstu = require('../js/index.js');

test('deriveSubkey is deterministic', () => {
  const masterKey = dstu.kdfKeygen();
  assert.deepStrictEqual(
    dstu.kdfDeriveSubkey(masterKey, 0, Buffer.from('encrypt_')),
    dstu.kdfDeriveSubkey(masterKey, 0, Buffer.from('encrypt_')),
  );
});

test('different subkeyId gives a different subkey', () => {
  const masterKey = dstu.kdfKeygen();
  const a = dstu.kdfDeriveSubkey(masterKey, 0, Buffer.from('context1'));
  const b = dstu.kdfDeriveSubkey(masterKey, 1, Buffer.from('context1'));
  assert.notDeepStrictEqual(a, b);
});

test('different context gives a different subkey', () => {
  const masterKey = dstu.kdfKeygen();
  const a = dstu.kdfDeriveSubkey(masterKey, 0, Buffer.from('context1'));
  const b = dstu.kdfDeriveSubkey(masterKey, 0, Buffer.from('context2'));
  assert.notDeepStrictEqual(a, b);
});

test('wrong-length master key is rejected', () => {
  assert.throws(() => dstu.kdfDeriveSubkey(Buffer.from('too short'), 0, Buffer.from('context1')));
});

test('wrong-length context is rejected', () => {
  const masterKey = dstu.kdfKeygen();
  assert.throws(() => dstu.kdfDeriveSubkey(masterKey, 0, Buffer.from('short')));
});

test('negative subkeyId is rejected', () => {
  const masterKey = dstu.kdfKeygen();
  assert.throws(() => dstu.kdfDeriveSubkey(masterKey, -1, Buffer.from('context1')));
});

File: bindings/go/dstu/kdf_test.go
package dstu

import (
	"bytes"
	"testing"
)

// Kdf (crypto_kdf) - no official vector exists (D-45). Correctness here means determinism/
// distinctness. Misuse: wrong-length master key/context.

func TestDeriveSubkeyIsDeterministic(t *testing.T) {
	key, _ := GenerateKdfMasterKey()
	defer key.Close()
	context := []byte("encrypt_")
	a, err := key.DeriveSubkey(0, context)
	if err != nil {
		t.Fatal(err)
	}
	b, err := key.DeriveSubkey(0, context)
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(a, b) {
		t.Fatal("expected the same subkey twice")
	}
}

func TestDifferentSubkeyIDGivesDifferentSubkey(t *testing.T) {
	key, _ := GenerateKdfMasterKey()
	defer key.Close()
	context := []byte("context1")
	a, _ := key.DeriveSubkey(0, context)
	b, _ := key.DeriveSubkey(1, context)
	if bytes.Equal(a, b) {
		t.Fatal("expected different subkeys")
	}
}

func TestDifferentContextGivesDifferentSubkey(t *testing.T) {
	key, _ := GenerateKdfMasterKey()
	defer key.Close()
	a, _ := key.DeriveSubkey(0, []byte("context1"))
	b, _ := key.DeriveSubkey(0, []byte("context2"))
	if bytes.Equal(a, b) {
		t.Fatal("expected different subkeys")
	}
}

func TestKdfWrongLengthMasterKeyIsRejected(t *testing.T) {
	if _, err := KdfMasterKeyFromBytes(make([]byte, 10)); err == nil {
		t.Fatal("expected an error")
	}
}

func TestKdfWrongLengthContextIsRejected(t *testing.T) {
	key, _ := GenerateKdfMasterKey()
	defer key.Close()
	if _, err := key.DeriveSubkey(0, []byte("short")); err == nil {
		t.Fatal("expected an error")
	}
}
